### PR TITLE
Dropdown renderer

### LINKED ISSUE #395: "Dropdown" question type: implement bare bones
This question type creates a dropdown with multiple choices available. Applicants may pick one or more choices.

image

If we have time, we can automatically add search bars atop long dropdown lists to help filter.

Make it possible for UAT admins to create, and for applicants to fill out, questions of this type. Answers should be saved to a database. Configuration options and validation criteria aren't necessary yet.

Context: https://github.com/seattle-uat/universal-application-tool/wiki/Question-type-requirements-notes

### PR DESCRIPTION
**Pull request by carolinedanzi**
### Description
1. Adds logic for rendering dropdown questions
2. Adds admin functionality for adding question options (note: this will not save to the database yet)
3. Adds a question option HTML template, which is copied and added in javascript each time the admin selects "Add answer option" button

### Checklist
- [x] Created tests which fail without the change (if possible)
- [ ] Extended the README / documentation, if necessary

### Issue(s)
Fixes #395
#631 

Screen Shot 2021-03-31 at 12 19 33 PM
Screen Shot 2021-03-31 at 12 23 02 PM

### CODE REVIEW

File: universal-application-tool-0.0.1/app/views/admin/questions/QuestionConfig.java [hunk 1]
```diff
@@ -105,6 +111,11 @@ private QuestionConfig addTextQuestionConfig() {
     return this;
   }
 
+  private QuestionConfig addMultiOptionQuestionConfig() {
+    content.with(button("Add answer option").withId("add-new-option").withClasses(Styles.M_2));
```
Comment (AWBaumann) [inline universal-application-tool-0.0.1/app/views/admin/questions/QuestionConfig.java:115]:
```suggestion
    content.with(button("Add answer option").withId("add-new-option").withClasses(Styles.M_2).withType("button"));
```

File: universal-application-tool-0.0.1/app/views/admin/questions/QuestionEditView.java [hunk 2]
```diff
@@ -103,7 +103,22 @@ private ContainerTag buildQuestionContainer(String title) {
...
+  // A hidden template for multi-option questions.
+  private ContainerTag multiOptionQuestionField() {
+    return div()
+        .with(
+            FieldWithLabel.input()
+                .setFieldName("option")
+                .setLabelText("Question option")
+                .getContainer()
+                .withClasses(Styles.FLEX, Styles.ML_2),
+            button("Remove").withClasses(Styles.FLEX, Styles.ML_4))
```
Comment (AWBaumann) [inline universal-application-tool-0.0.1/app/views/admin/questions/QuestionEditView.java:119]:
```suggestion
            button("Remove").withClasses(Styles.FLEX, Styles.ML_4).withType("button"))
```

File: universal-application-tool-0.0.1/app/assets/javascripts/main.ts [hunk 3]
```diff
@@ -63,10 +63,44 @@ function dismissWarning() {
   localStorage.setItem("hideWarning", "true");
 }
 
+/** In the admin question form - add a new option input for each new question answer option. */
+function addNewQuestionAnswerOptionForm(event: Event) {
+  // Prevent the default event, which is to submit the containing form.
+  event.preventDefault();
```
[the remark this answers is not in the document]
Comment (bion) [reply universal-application-tool-0.0.1/app/assets/javascripts/main.ts:69]: Ah this is good to know, I suggested `preventDefault`

[the remark this answers is not in the document]
Comment (carolinedanzi) [reply universal-application-tool-0.0.1/app/assets/javascripts/main.ts:69]: Amazing! Learned something new today

File: universal-application-tool-0.0.1/app/views/admin/questions/QuestionConfig.java (hunk 1 above)
Comment (carolinedanzi) [reply universal-application-tool-0.0.1/app/views/admin/questions/QuestionConfig.java]: Done!

File: universal-application-tool-0.0.1/app/views/admin/questions/QuestionEditView.java (hunk 2 above)
Comment (carolinedanzi) [reply universal-application-tool-0.0.1/app/views/admin/questions/QuestionEditView.java]: Done!

File: universal-application-tool-0.0.1/app/assets/javascripts/main.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (carolinedanzi) [reply universal-application-tool-0.0.1/app/assets/javascripts/main.ts]: Updated!

File: browser-test/src/admin_program_creation.test.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    await page.click('button:text("Remove"):visible')
+
+    // Assert there are only two options now
+    questionSettingsDiv = await page.innerHTML('#question-settings')
+    expect(questionSettingsDiv.match(/<input/g).length).toBe(2)
+
+    // TODO(https://github.com/seattle-uat/civiform/issues/631):
+    // Assert options are written correctly and appear once form is submitted
+  })
+})
```
Comment (AWBaumann) [inline browser-test/src/admin_program_creation.test.ts:35]: Add new line at end of file.

Also, this looks so much better than Selenium.

File: universal-application-tool-0.0.1/app/views/admin/questions/QuestionEditView.java [hunk 5]
```diff
@@ -103,7 +103,22 @@ private ContainerTag buildQuestionContainer(String title) {
...
+    return div()
+        .with(
+            FieldWithLabel.input()
+                .setFieldName("option")
+                .setLabelText("Question option")
+                .getContainer()
+                .withClasses(Styles.FLEX, Styles.ML_2),
+            button("Remove").withType("button").withClasses(Styles.FLEX, Styles.ML_4))
+        .withId("multi-option-question-answer-template")
+        .withClasses(Styles.HIDDEN, Styles.FLEX, Styles.FLEX_ROW, Styles.MB_4);
```
Comment (AWBaumann) [inline universal-application-tool-0.0.1/app/views/admin/questions/QuestionEditView.java:121]: Nit. Consider defining the input and remove button outside of the return statement for more readable code...

```suggestion
    ContainerTag optionInput = FieldWithLabel.input()
                .setFieldName("option")
                .setLabelText("Question option")
                .getContainer()
                .withClasses(Styles.FLEX, Styles.ML_2);
    ContainerTag removeOptionButton = 
            button("Remove").withType("button").withClasses(Styles.FLEX, Styles.ML_4);
    return div(optionInput, removeOptionButton)
        .withId("multi-option-question-answer-template")
        .withClasses(Styles.HIDDEN, Styles.FLEX, Styles.FLEX_ROW, Styles.MB_4);
```
or


```suggestion
    ContainerTag optionInput = FieldWithLabel.input()
                .setFieldName("option")
                .setLabelText("Question option")
                .getContainer()
                .withClasses(Styles.FLEX, Styles.ML_2);
    ContainerTag removeOptionButton = 
            button("Remove").withType("button").withClasses(Styles.FLEX, Styles.ML_4);
    return div()
        .withId("multi-option-question-answer-template")
        .withClasses(Styles.HIDDEN, Styles.FLEX, Styles.FLEX_ROW, Styles.MB_4)
        .with(optionInput, optionButton);
```

**Review by AWBaumann — APPROVED** [review]

File: universal-application-tool-0.0.1/app/views/admin/questions/QuestionEditView.java (hunk 5 above)
Comment (carolinedanzi) [reply universal-application-tool-0.0.1/app/views/admin/questions/QuestionEditView.java]: Much more readable, thank you!

File: browser-test/src/admin_program_creation.test.ts (hunk 4 above)
Comment (carolinedanzi) [reply browser-test/src/admin_program_creation.test.ts]: Done!

**Pull request merged**